streaming/storage/deque.go

```diff
@@ -0,0 +1,383 @@
+}
+
+func (ll *Deque) appendBytes(data []byte, index int, modifier func() error) error {
+	byteKey := getIndexKey(index)
+
+	err := ll.tx.Set(byteKey, data)
+	if err != nil {
+		return errors.Wrap(err, "couldn't add the element to the queue")
+	}
+
+	err = modifier()
+	if err != nil {
+		return errors.Wrap(err, "failed to actualize queue attributes")
+	}
+
+	return nil
+}
+
+//Removes the first element from the queue and returns it
+//Returns ErrEmptyQueue if the list is empty
+func (ll *Deque) PopFront(value proto.Message) error {
+	//there is no need to call initialize here, since PeekFront calls initialize
+	err := ll.PeekFront(value)
+	if err != nil {
+		return err
+	}
+
+	return ll.pop(value, ll.firstElement+1, ll.incFirst)
+}
+
+//Removes the last element from the queue and returns it
+//Returns ErrEmptyQueue if the list is empty
+func (ll *Deque) PopBack(value proto.Message) error {
+	//there is no need to call initialize here, since PeekFront calls initialize
+	err := ll.PeekBack(value)
+	if err != nil {
+		return err
+	}
+
+	return ll.pop(value, ll.lastElement-1, ll.decLast)
+}
+
+func (ll *Deque) pop(value proto.Message, index int, modifier func() error) error {
+	err := ll.peekIndex(value, index)
+	if err != nil {
+		return err
+	}
+
+	key := getIndexKey(index)
+
+	err = ll.tx.Delete(key)
+	if err != nil {
+		return errors.New("failed to delete element from queue in pop")
+	}
+
+	err = modifier()
+	return errors.Wrap(err, "failed to actualize queue attributes in pop")
+}
+
+//Returns the first element of the queue without removing it
+//Returns ErrEmptyQueue if the queue is empty
+func (ll *Deque) PeekFront(value proto.Message) error {
+	if !ll.initialized {
+		err := ll.initialize()
+		if err != nil {
+			return errors.Wrap(err, "failed to initialize in queue.PeekFront()")
+		}
+	}
+
+	return ll.peekIndex(value, ll.firstElement+1)
+}
+
+//Returns the last element of the queue without removing it
+//Returns ErrEmptyQueue if the queue is empty
+func (ll *Deque) PeekBack(value proto.Message) error {
+	if !ll.initialized {
+		err := ll.initialize()
+		if err != nil {
+			return errors.Wrap(err, "failed to initialize in queue.PeekBack()")
+		}
+	}
+
+	return ll.peekIndex(value, ll.lastElement-1)
+}
+
+//we don't check if the queue has been initialized here, since we only
+//call peekIndex from PeekFront/PeekBack, which both initialize the queue
+func (ll *Deque) peekIndex(value proto.Message, index int) error {
+	if ll.lastElement-ll.firstElement <= 1 {
+		return ErrEmptyQueue
+	}
+
+	key := getIndexKey(index)
+
+	data, err := ll.tx.Get(key)
+	if err != nil {
+		return errors.New("couldn't get value of element in peek")
+	}
+
+	err = proto.Unmarshal(data, value)
+	return errors.Wrap(err, "couldn't unmarshal the element in peek")
+}
+
+func (ll *Deque) Print() error { //add initialize here (?)
+	var value octosql.Value
+	it := ll.tx.WithPrefix(dequeValueKeyPrefix).Iterator()
+
+	defer func() {
+		_ = it.Close() //TODO: panic here?
```

Comment:
why would you panic? :D

Reply:
This was a debug function I forgot to remove, sorry about that :3